src/object/srv_ec_aggregate.c

```diff
@@ -560,31 +806,779 @@ agg_update_vos(struct ec_agg_entry *entry, bool write_parity)
-		if (erc)
-			D_ERROR("vos_obj_array_remove failed: "DF_RC"\n",
-				DP_RC(erc));
-		if (!rc && erc)
-			rc = erc;
-		ext_counter++;
-		if (ext_counter >= ext_total)
-			break;
+		D_ASSERT(!extent->ae_hole);
+		if (extent->ae_epoch <= entry->ae_par_extent.ape_epoch)
+			continue;
+		if (estart == extent->ae_recx.rx_idx - ss) {
+			eend = estart + extent->ae_recx.rx_nr;
+			elen = extent->ae_recx.rx_nr;
+			continue;
+		}
+		if (extent->ae_recx.rx_idx - ss > eend) {
+			cell_cnt += agg_count_cells(fcbit_map, tbit_map, estart,
+						    elen, k, len, entry->
+						    ae_cur_stripe.as_stripenum,
+						    &full_cell_cnt);
+			estart = extent->ae_recx.rx_idx;
+			elen = 0;
+		}
+		elen += extent->ae_recx.rx_nr;
+		eend += extent->ae_recx.rx_nr;
+	}
+	cell_cnt += agg_count_cells(fcbit_map, tbit_map, estart, elen, k, len,
+				    entry->ae_cur_stripe.as_stripenum,
+				    &full_cell_cnt);
+
+	if (full_cell_cnt >= k / 2 || cell_cnt == k) {
+		stripe_ud.asu_recalc = true;
+		cell_cnt = full_cell_cnt;
+		bit_map = fcbit_map;
+	} else
+		bit_map = tbit_map;
+
+	D_ASSERT(cell_cnt);
+	rc = agg_prep_sgl(entry);
+	if (rc)
+		goto out;
+	if (cell_cnt)
+		/* cell_cnt is zero if all cells are partial filled */
+		rc = agg_fetch_local_extents(entry, bit_map, cell_cnt,
+					     stripe_ud.asu_recalc);
+	if (rc)
+		goto out;
+
+	if (stripe_ud.asu_recalc) {
+		for (i = 0; i < k; i++) {
+			if (isset(bit_map, i))
+				clrbit(bit_map, i);
+			else
+				setbit(bit_map, i);
+		}
+	}
+
+	stripe_ud.asu_agg_entry = entry;
+	stripe_ud.asu_bit_map = bit_map;
+	stripe_ud.asu_cell_cnt = cell_cnt;
+
+	rc = ABT_eventual_create(sizeof(*status), &stripe_ud.asu_eventual);
+	if (rc != ABT_SUCCESS) {
+		rc = dss_abterr2der(rc);
+		goto out;
+	}
+	rc = dss_ult_create(agg_process_partial_stripe_ult, &stripe_ud,
+			    DSS_ULT_EC, 0, 0, NULL);
+	if (rc)
+		goto ev_out;
+	rc = ABT_eventual_wait(stripe_ud.asu_eventual, (void **)&status);
+	if (rc != ABT_SUCCESS) {
+		rc = dss_abterr2der(rc);
+		goto ev_out;
+	}
+	if (*status != 0) {
+		rc = *status;
+		goto ev_out;
 	}
+
+ev_out:
+	ABT_eventual_free(&stripe_ud.asu_eventual);
+
+out:
+	return rc;
+}
+
+/* Sends the generated parity and the stripe number to the peer
+ * parity target. Handler writes the parity and deletes the replicas
+ * for the stripe.  Has to be extended to support p > 2.
+ */
+static void
+agg_peer_update_ult(void *arg)
+{
+	struct ec_agg_stripe_ud	*stripe_ud = (struct ec_agg_stripe_ud *)arg;
+	struct ec_agg_entry	*entry = stripe_ud->asu_agg_entry;
+	d_iov_t			 iov = { 0 };
+	d_sg_list_t		 sgl = { 0 };
+	crt_endpoint_t		 tgt_ep = { 0 };
+	unsigned char		*buf = NULL;
+	struct obj_ec_agg_in	*ec_agg_in = NULL;
+	struct obj_ec_agg_out	*ec_agg_out = NULL;
+	struct ec_agg_param	*agg_param;
+	crt_opcode_t		 opcode;
+	crt_rpc_t		*rpc;
+	int			 rc = 0;
+
+	tgt_ep.ep_rank = entry->ae_peer_rank;
+	tgt_ep.ep_tag = entry->ae_peer_idx + 1;
+	opcode = DAOS_RPC_OPCODE(DAOS_OBJ_RPC_EC_AGGREGATE, DAOS_OBJ_MODULE,
```

Comment:
you probably need check if the peer is still active in pool map? if not, it does not need send rpc to the peer.   Otherwise it will cause unnecessary aggregation failure for degraded mode.

Reply:
I know this is an issue.  How should I do this?  Thanks.

Reply:
oh, you can use  this function to retrieve all of failure targets. 

 rc = pool_map_find_failed_tgts(pool->sp_map, &targets,
                                               &failed_tgts_cnt);
if you need know rank, you can check rebuild_leader_status_check

Reply:
Thanks, I will add this check.

Reply:
I check it at the top of agg_process_stripe(), since any missing target will potentially prevent aggregating the stripe.